File: backend/src/main/kotlin/ai/project/rio/transaction/Transaction.kt
package ai.project.rio.transaction

import ai.project.rio.money.Money
import java.time.Instant

enum class TransactionType { CREDIT, DEBIT }

enum class TransactionStatus { PENDING, COMPLETED, DECLINED }

/**
 * `amount` is a magnitude and is always > 0 for persisted transactions.
 * Direction (money in vs. money out) comes from `type`, never from the sign of the amount.
 */
data class Transaction(
    val id: String,
    val description: String,
    val amount: Money,
    val type: TransactionType,
    val status: TransactionStatus,
    val createdAt: Instant,
)

File: frontend/src/types/transaction.ts
import type { Money } from "../money/money";

// Domain types used by pages and components. Money is already converted to bigint here.

export type TransactionType = "CREDIT" | "DEBIT";

export type TransactionStatus = "PENDING" | "COMPLETED" | "DECLINED";

export interface Transaction {
  id: string;
  description: string;
  /** Magnitude, always > 0. Direction comes from `type`. */
  amount: Money;
  type: TransactionType;
  status: TransactionStatus;
  createdAt: Date;
  /** The validated API instant, verbatim (e.g. `2026-09-02T15:30:00.123456789Z`). Date truncates to milliseconds; use this for exact display and log correlation. */
  createdAtInstant: string;
}

export interface CreateTransactionInput {
  description: string;
  amount: Money;
  type: TransactionType;
}
